### PR TITLE
[DataGrid] Fix quick filter input lag

### PR DESCRIPTION
**Pull request by cherniavskii**
- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Fixes https://github.com/mui/mui-x/issues/6783

Before: https://codesandbox.io/s/infallible-villani-gmwpg6
After: https://codesandbox.io/s/long-tree-q7p8c9

### CODE REVIEW

File: packages/grid/x-data-grid/src/tests/quickFiltering.DataGrid.test.tsx [hunk 1]
```diff
@@ -372,4 +400,40 @@ describe('<DataGrid /> - Quick Filter', () => {
       expect(getRows({ quickFilterValues: [''] }).year).to.deep.equal(ALL_ROWS_YEAR);
     });
   });
+
+  // See https://github.com/mui/mui-x/issues/6783
+  it('should not override user input when typing', async function test() {
+    if (isJSDOM) {
+      this.skip();
+    }
+    // Warning: this test doesn't fail consistently as it is timing-sensitive.
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/tests/quickFiltering.DataGrid.test.tsx:409]: While this test doesn't consistently fail without the fix, I thought it was still better than nothing 🙂
What do you think?

File: packages/grid/x-data-grid/src/components/toolbar/GridToolbarQuickFilter.tsx [hunk 2]
```diff
@@ -59,7 +59,7 @@ export type GridToolbarQuickFilterProps = TextFieldProps & {
    * @param {any[]} values The new values passed to the quick filter model
    * @returns {string} The string to display in the text field
    */
-  quickFilterFormatter?: (values: GridFilterModel['quickFilterValues']) => string;
+  quickFilterFormatter?: (values: NonNullable<GridFilterModel['quickFilterValues']>) => string;
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/components/toolbar/GridToolbarQuickFilter.tsx:62]: We never pass `undefined` as `values`, so we should exclude it here

**Review by romgrk — APPROVED** [review]

File: packages/grid/x-data-grid/src/tests/quickFiltering.DataGrid.test.tsx (hunk 1 above)
Comment (romgrk) [reply packages/grid/x-data-grid/src/tests/quickFiltering.DataGrid.test.tsx:409]: Not ideal but it's fine as it is.

File: packages/grid/x-data-grid/src/components/toolbar/GridToolbarQuickFilter.tsx [hunk 3]
```diff
@@ -101,7 +101,9 @@ function GridToolbarQuickFilter(props: GridToolbarQuickFilterProps) {
 
   const updateSearchValue = React.useCallback(
     (newSearchValue: string) => {
-      apiRef.current.setQuickFilterValues(quickFilterParser(newSearchValue));
+      const newQuickFilterValues = quickFilterParser(newSearchValue);
+      setPrevQuickFilterValues(newQuickFilterValues);
```
[the remark this answers is not in the document]
Comment (cherniavskii) [reply packages/grid/x-data-grid/src/components/toolbar/GridToolbarQuickFilter.tsx:105]: Good point, it should be a ref indeed!

**Pull request merged**